### ISSUE TITLE
load custom driver without steamvr

### ISSUE THREAD

**Issue by SlimSnow** [ValveSoftware/openvr#1157]
I want to load driver with is develop by myself without steamvr.I noticed that there is a entry point in the driver code like "HmdDriverFactory".And I get this with "GetProcAddress",But I can not initialize my driver.Who can tell me how to initialize my driver in my own code without steamvr.

**Comment by TheDeveloperGuy** [comment]
A SteamVR driver without SteamVR is not very useful. What are you trying to achieve exactly? The driver interface is a fairly complex interaction of C++ objects, most of which don't live in your driver. So, whatever your app is, it would have to emulate big portions of OpenVR to use the driver. If you have some unusual use case where you need to use portions of your driver outside of SteamVR, then it would be a lot easier to provide your own entry points to that functionality.

**Comment by SlimSnow** [comment]
@TheDeveloperGuy SteamVR have many functions,but I use SteamVR just want to load my driver.And In issue #291 I notice that OpenVR is the API. SteamVR is the implementation of that API.So I want to develop my own "SteamVR".

**Comment by TheDeveloperGuy** [comment]
No offence, but if you're asking this question, I suspect you lack the knowledge/experience to develop your own "SteamVR". Nonetheless, you can look at OpenComposite, a non-SteamVR implementation of OpenVR.

**Comment by SlimSnow** [comment]
@TheDeveloperGuy I konw that,I just want to load driver without SteamVR.Thank you for your advise.

**Comment by risa2000** [comment]
@SlimSnow I did the exercise few months back, but realized that in order to _initialize_ the driver, the upper layer (whatever it is) must pass to it a bunch of objects, while potentially cross referencing the driver objects, and it is a mess. I am not sure OpenComposite will help you much (just from the look) as it seems to emulate the application layer, but not the driver interface, and I would assume the internals are just the wrapper calls to OVR.

So, in short, the easiest way to initialize an OpenVR driver is to plug it in OpenVR.

**Comment by SlimSnow** [comment]
@risa2000 Thank you for your advise.

**Comment by spayne** [comment]
@SlimSnow I've also wanted this ability to load initialize a driver.  For me it's been for testing purposes:

- I wanted in one case to automatically regression test: Write the driver once and be able to automatically verify that with every device update and with every openvr update that everything is working together properly.  But this auto regression case doesn't look promising because I'm wanting the system to be very conformant to the openvr updates and so it becomes necessary to be an expert in openvr's behaviour and so at that point the best approach becomes to test against openvr itself (as @risa2000  said:  So, in short, the easiest way to initialize an OpenVR driver is to plug it in OpenVR)
- In another case I wanted to verify some corner cases around power management.  But again, for me the system would need to be very conformant to how OpenVr was going to act.
- In a third case I wanted assurance that the driver was correct.  e.g. is there a set of properties that all drivers must implement. OpenVR isn't tightly defined though - so it's basically back to defacto testing.

I've looked around a bit at what other systems do.  For example, Windows (https://docs.microsoft.com/en-us/windows-hardware/drivers/develop/testing-a-driver) has ways to remotely launch and debug an OS on test hardware.  ie there is an emphasis here on actual live system testing of both the os and the device.  The driver writer's job to be an expert in the host and the device and write the probably unsexy and probably 'hardware engineery' mappings between them.    

OpenVR is broader than most people expect - and writing alot of wrappers for that would be annoying and very possible - but that's not the actual issue: the thing I want is the exact behaviour of openvr - so the return on time investment hasn't been justifiable yet. 

I know I'm (probably) echoing what @risa2000 and @TheDeveloperGuy  are saying.  I'm hoping by adding some details as well that is also helpful.

**Comment by JoeLudwig** [comment]
Hi @SlimSnow 

Since our implementation is SteamVR, we obviously don't support running a driver without it. So I'm closing this issue.
